jbake/content/authentication.adoc

```diff
@@ -0,0 +1,389 @@
+
+But what happens if the login attempt failed?
+For example, what if the end-user supplied an incorrect password, or accessed the system too many times and maybe their account is locked?
+
+Shiro has a rich runtime link:/static/current/apidocs/org/apache/shiro/authc/AuthenticationException.html[`AuthenticationException`] hierarchy that can indicate exactly why the attempt failed.
+You can wrap `login` in a `try/catch` block and catch any exception you wish and react to them accordingly.
+For example:
+
+[source,java]
+----
+try {
+    currentUser.login(token);
+} catch ( UnknownAccountException uae ) { ...
+} catch ( IncorrectCredentialsException ice ) { ...
+} catch ( LockedAccountException lae ) { ...
+} catch ( ExcessiveAttemptsException eae ) { ...
+} ... catch your own ...
+} catch ( AuthenticationException ae ) {
+    //unexpected error?
+}
+
+//No problems, continue on as expected...
+----
+
+If one of the existing exception classes do not meet your needs, custom `AuthenticationExceptions` can be created to represent specific failure scenarios.
+
+[TIP]
+====
+.Login Failure Tip
+While your code can react to specific exceptions and execute logic as necessary, a security best practice is to only show a generic failure message to an end user in the event of a failure, for example, "Incorrect username or password.".
+This ensures no specific information is available to hackers that may be attempting an attack vector.
+====
+
+[#Authentication-RememberedVsAuthenticated]
+== Remembered vs. Authenticated
+
+As shown in the example above, Shiro supports the notion of "remember me" in addition to the normal login process.
+It is worth pointing out at this time that Shiro makes a very precise distinction between a _remembered_ Subject and an actual _authenticated_ Subject:
+
+* *Remembered*: A remembered `Subject` is not anonymous and has a known identity (i.e. `subject.link:/static/current/apidocs/org/apache/shiro/subject/Subject.html#getPrincipals()[getPrincipals()]` is non-empty).
+But this identity is remembered from a previous authentication during a *previous* session.
+A subject is considered remembered if `subject.link:/static/current/apidocs/org/apache/shiro/subject/Subject.html#isRemembered()[isRemembered()]` returns `true`.
+
+* *Authenticated*: An authenticated `Subject` is one that has been successfully authenticated (i.e. the `login` method was invoked without throwing an exception) _during the Subject's current session_.
+A subject is considered authenticated if `subject.link:/static/current/apidocs/org/apache/shiro/subject/Subject.html#isAuthenticated()[isAuthenticated()]` returns `true`.
+
+[WARNING]
+====
+.Mutually Exclusive
+Remembered and authenticated states are mutually exclusive - a `true` value for one indicates a `false` value for the other and vice versa.
+====
```

Comment:
According to https://asciidoctor.org/docs/asciidoc-writers-guide/#admonitions, I might argue that  `CAUTION` would be the better choice (now that we have it available):

* Use CAUTION to advise the reader to act carefully (i.e., exercise care).
* Use WARNING to inform the reader of danger, harm, or consequences that exist.